### PR TITLE
collective selection (cleaned)

### PR DESCRIPTION
**Pull request by raffenet**

### CODE REVIEW

**Comment by raffenet** [comment]
@AlexeyMalkhanov I have created this new PR based on some cleanups and re-ordering I did while review. Can you please have a look at it and see if it matches your expectations?

File: src/mpid/ch4/src/ch4_coll.h [hunk 1]
```diff
@@ -13,49 +13,455 @@
...
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_Allreduce_composition_alpha(const void *sendbuf, void *recvbuf, int count,
+                                                               MPI_Datatype datatype, MPI_Op op, MPIR_Comm * comm,
+                                                               MPIR_Errflag_t * errflag,
+                                                               MPIDI_coll_algo_container_t * ch4_algo_parameters_container)
+{
+    int mpi_errno = MPI_SUCCESS;
+    void * reduce_node_container = MPIDI_coll_get_next_container(ch4_algo_parameters_container);
+    void * allred_roots_container = MPIDI_coll_get_next_container(reduce_node_container);
+    void * bcast_node_container = MPIDI_coll_get_next_container(bcast_node_container);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_coll.h]: The compiler is complaining that `bcast_node_container` is used uninitialized here.
```
src/mpi/coll/allreduce.c: In function ‘MPIDI_Allreduce_composition_alpha’:
./src/mpid/ch4/src/ch4_coll.h:353:12: warning: ‘bcast_node_container’ is used uninitialized in this function [-Wuninitialized]
```

Comment (raffenet) [inline src/mpid/ch4/src/ch4_coll.h]: @AlexeyMalkhanov

Comment (AlexeyMalkhanov) [inline src/mpid/ch4/src/ch4_coll.h]: [PR with fixes](https://github.com/raffenet/mpich/pull/4)

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've created the PR to this PR which address the mentioned issues. I've looked through the new PR and it seems to be exactly what we planned.

**Comment by raffenet** [comment]
@AlexeyMalkhanov I have integrated your changes into this PR, as well as squashed my cleanup commit. Re-running the tests.
test:jenkins/ch4/most

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've created one more PR with file structures reorganizations. [file dependency resolution](https://github.com/raffenet/mpich/pull/5)

**Comment by raffenet** [comment]
OK, I will include it here and re-run the tests.

**Comment by raffenet** [comment]
This looks mostly OK. There are some new failures in the AM-only configuration related to comm creation, so I suspect there may be an allreduce problem in there somewhere. Since AM-only is not default, and we don't even have a working AM-only config for ch4/ofi, I won't block on that for now. But we should fix that eventually.

**Comment by raffenet** [comment]
Just doing a sanity check on ch3. test:jenkins/ch3/tcp

**Comment by AlexeyMalkhanov** [comment]
@raffenet, definitely that should be fixed.  Are there a chance to have the AM-only configuration that we would able to reproduce the issue and possibly fix it?

**Comment by raffenet** [comment]
@AlexeyMalkhanov with ch4/ofi, you just need to set these envvars to force everything over AMs.

  

>  MPIR_CVAR_CH4_OFI_ENABLE_TAGGED=0
>     export MPIR_CVAR_CH4_OFI_ENABLE_TAGGED
>     MPIR_CVAR_CH4_OFI_ENABLE_RMA=0
>     export MPIR_CVAR_CH4_OFI_ENABLE_RMA
>

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've checked the master branch from https://github.com/pmodels/mpich.git and behavior is the same as with our changes: some tests fail with the envvars specified.

**Comment by raffenet** [comment]
I think we are ready. I squashed things down a little more and signed off. Merging...

**Pull request merged**